File: ABC/ABC143/E.py
import sys

sys.setrecursionlimit(10 ** 7)
rl = sys.stdin.readline


def solve():
    N, M, L = map(int, rl().split())
    INF = 10 ** 18
    dist = [[INF] * N for _ in range(N)]
    for _ in range(M):
        a, b, c = map(int, rl().split())
        a, b = a - 1, b - 1
        dist[a][b] = dist[b][a] = c
    
    for k in range(N):
        for i in range(N):
            for j in range(N):
                dist[i][j] = min(dist[i][j], dist[i][k] + dist[k][j])
    
    cost = [[INF] * N for _ in range(N)]
    for i in range(N):
        for j in range(N):
            if dist[i][j] <= L:
                cost[i][j] = 1
    
    for k in range(N):
        for i in range(N):
            for j in range(N):
                cost[i][j] = min(cost[i][j], cost[i][k] + cost[k][j])
    
    Q = int(rl())
    st = [list(map(lambda n: int(n) - 1, rl().split())) for _ in range(Q)]
    ans = [-1] * Q
    for i, (s, t) in enumerate(st):
        if cost[s][t] != INF:
            ans[i] = cost[s][t] - 1
    print(*ans, sep='\n')


if __name__ == '__main__':
    solve()

File: ABC/ABC143/E.nim
import math, sequtils, strutils


type
    Graph[N: static[int]; T] = array[N, array[N, T]]

proc floyd_warshall*(graph: var Graph, n: Natural) =
    for k in 0..<n:
        for i in 0..<n:
            for j in 0..<n:
                graph[i][j] = min(graph[i][j], graph[i][k] + graph[k][j])


const INF = 10^18
var dist, cost: Graph[300, int]


proc solve() =
    var N, M, L, a, b, c: int
    (N, M, L) = stdin.readLine.split.map(parseInt)
    for i in 0..<N:
        for j in 0..<N:
            dist[i][j] = INF
            cost[i][j] = INF
    for _ in 0..<M:
        (a, b, c) = stdin.readLine.split.map(parseInt)
        dec a; dec b
        dist[a][b] = c
        dist[b][a] = c
    
    dist.floyd_warshall(N)
    
    for i in 0..<N:
        for j in 0..<N:
            if dist[i][j] <= L:
                cost[i][j] = 1
    
    cost.floyd_warshall(N)
    
    var
        Q = stdin.readLine.parseInt
        ans = newSeqWith(Q, -1)
        s, t: int
    for i in 0..<Q:
        (s, t) = stdin.readLine.split.mapIt(it.parseInt - 1)
        if cost[s][t] != INF:
            ans[i] = cost[s][t] - 1
    echo ans.join("\n")


when is_main_module:
    solve()
